aten/src/ATen/native/cpu/Reduce.h

```diff
@@ -154,12 +154,13 @@ struct all_same : c10::guts::conjunction<
 // data_t is the input/output data type.
 // acc_t is a type that contains all the necessary data
 // to continue reducing.
+// index_t is a one-dimensional index
 //
 // ops_t is such that &ops_t::reduce, &ops_t::combine, and &ops_t::project exist and satisfy
 // the following.
-// reduce: (acc_t, data_t) -> acc_t adds one data point to the accumulated value.
+// reduce: (acc_t, data_t, index_t) -> acc_t adds one data point to the accumulated value.
```

Comment:
So now the signatures of all reduction functors include index_t argument, even those that don't need it? This seems not ideal. Is it possible to make acc_t hold both value and index, or better yet, only index (for argmin/argmax you need only index, right?)

Reply:
>Is it possible to make acc_t hold both value and index

This is exactly what `ArgMaxOps` does. `acc_t` is a black box as far as the generic reduction logic is concerned. It's only the reduction `ops_t` which can add data members to it, so the `ops_t` needs access to the current index.

>or better yet, only index (for argmin/argmax you need only index, right?)

Not true, you need the current maximum/minimum value to compare the current element against.

Reply:
You are right, just index is not enough.